Rough draft of scotty JSON for distributions.

keep94 opened the pull request:
Here is a strawman of how scotty will present distribution metrics it has collected in json.  Scotty will report distribution values only when they change just as scotty reports scalar values only when they change. 

In this PR, scotty provides only delta distribution statistics each time. For example, if scotty collects data once every 5 minutes, it will provide distribution statistics in 5 minute intervals.  If an endpoint goes down and is brought back up between collections, scotty will detect this with the help of the generationId field and will show no distribution for that particular interval.

Clients of scotty will aggregate distribution times for themselves. For instance, if a client wants RPC latency between 11:00 and 12:00, it will ask scotty for all the delta distributions between 11:00 and 12:00 and aggregate for itself. This gives the client the freedom to deal with endpoint restarts how it chooses.  For instance, if a restart happened between 11:00 and 12:00, the client could ignore it and just approximate the distribution for that hour, or the client could build one distribution before the restart and one after the restart.

keep94 committed:
Rough draft of scotty JSON for distributions.

keep94 committed:
Explain that scotty stores zero distributions for a restarted endpoint

rgooch commented on messages/api.go:
```diff
@@ -6,6 +6,37 @@ import (
+	// Represents the upper bound of the range exclusive.
+	// Ignore for the highest (last) range which never has a upper bound.
+	Upper float64 `json:"upper"`
+}
```
I see later on that you define buckets using []Range. I would expect that range[n].Upper == range[n+1].Lower, so there appears to be duplication of data. Why not just have []float64 for the bucket endpoints. You would have B+1 endpoints, where B = number of buckets.

keep94 replied:
Done. Thanks.

keep94 committed:
Richard's suggestion.

rgooch commented on messages/api.go:
```diff
@@ -78,8 +68,8 @@ type EndpointMetric struct {
-	Buckets []Range `json:"buckets,omitempty"`
+	// The upper limits for the distribution buckets.
+	UpperLimits []float64 `json:"upperLimits,omitempty"`
 }
```
How do you represent the lower limit for the first bucket?

keep94 replied:
In my distributions, the first bucket never has a lower limit. My buckets are always something like:
<5, 5-10, 10-20, 20-50, >50.

So -Inf always goes in the first bucket.

keep94 replied:
Also the last bucket never has an upper limit. So +Inf always goes in the last bucket.

rgooch commented:
OK. LGTM.

keep94 committed:
Not done.

Still have to code distribution extraction functions.
Still have to code up distribution difference algorithm.
Still got to write tests.
